std/algorithm.d

```diff
@@ -1409,6 +1412,149 @@ unittest
+/**
+Eagerly iterates over $(D r) and calls $(D pred) over _each element.
+
+Params:
+    pred = predicate to apply to each element of the range
+    r = range or iterable over which each iterates
+
+Example:
+---
+void deleteOldBackups()
+{
+    import std.algorithm, std.datetime, std.file;
+    auto cutoff = Clock.currTime() - 7.days;
+    dirEntries("", "*~", SpanMode.depth)
+        .filter!(de => de.timeLastModified < cutoff)
+        .each!remove();
+}
+---
+
+If the range supports it, the value can be mutated in place. Examples:
+---
+arr.each!((ref a) => a++);
+arr.each!"a++";
+---
+
+If no predicate is specified, $(D each) will default to doing nothing
+but consuming the entire range. $(D .front) will be evaluated, but this
+can be avoided by explicitly specifying a predicate lambda with a
+$(D lazy) parameter.
+
+$(D each) also supports $(D opApply)-based iterators, so it will work
+with e.g. $(XREF parallelism, parallel).
+
+See_Also: $(XREF range,tee)
+
+ */
+template each(alias pred = "a")
+{
+    alias BinaryArgs = TypeTuple!(pred, "i", "a");
+
+    enum isRangeUnaryIterable(R) =
+        is(typeof(unaryFun!pred(R.init.front)));
+
+    enum isRangeBinaryIterable(R) =
+        is(typeof(binaryFun!BinaryArgs(0, R.init.front)));
+
+    enum isRangeIterable(R) =
+        isInputRange!R &&
+        (isRangeUnaryIterable!R || isRangeBinaryIterable!R);
+
+    enum isForeachUnaryIterable(R) =
+        is(typeof((R r) {
+            foreach (ref a; r)
+                cast(void)unaryFun!pred(a);
+        }));
+
+    enum isForeachBinaryIterable(R) =
+        is(typeof((R r) {
+            foreach (i, ref a; r)
+                cast(void)binaryFun!BinaryArgs(i, a);
+        }));
+
+    enum isForeachIterable(R) =
+        (!isForwardRange!R || isDynamicArray!R) &&
+        (isForeachUnaryIterable!R || isForeachBinaryIterable!R);
+
+    void each(Range)(Range r)
+    if (isRangeIterable!Range && !isForeachIterable!Range)
+    {
+        debug(each) pragma(msg, "Using while for ", Range.stringof);
+        static if (isRangeUnaryIterable!Range)
+        {
+            while (!r.empty)
+            {
+                cast(void)unaryFun!pred(r.front);
+                r.popFront();
+            }
+        }
+        else // if (isRangeBinaryIterable!Range)
+        {
+            size_t i = 0;
+            while (!r.empty)
+            {
+                cast(void)binaryFun!BinaryArgs(i, r.front);
+                r.popFront();
+                i++;
+            }
+        }
+    }
+
+    void each(Iterable)(Iterable r)
+        if (isForeachIterable!Iterable)
+    {
+        debug(each) pragma(msg, "Using foreach for ", Iterable.stringof);
+        static if (isForeachUnaryIterable!Iterable)
+        {
+            foreach (ref e; r)
+                cast(void)unaryFun!pred(e);
+        }
+        else // if (isForeachBinaryIterable!Iterable)
+        {
+            foreach (i, ref e; r)
+                cast(void)binaryFun!BinaryArgs(i, e);
```

Comment:
The warning is useless because it does not give any context - it only points inside std.algorithm, without showing where `each` is instantiated.

There is already a unit test for this, as the warning manifests when removing the casts.